{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import importlib\n",
    "import json\n",
    "\n",
    "import experiment_runner\n",
    "import generate_synthetic_data\n",
    "import neural_network\n",
    "from generate_synthetic_data import GenerateSyntheticData\n",
    "from experiment_runner import ExperimentRunner\n",
    "importlib.reload(experiment_runner)\n",
    "importlib.reload(generate_synthetic_data)\n",
    "importlib.reload(neural_network)\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "# Import preprocessor\n",
    "from preprocess_data import PreprocessData"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Steps\n",
    "1. Import the files\n",
    "2. Transform all their features\n",
    "3. Use the experiment runner and the generate synthetic data class to get info for the experiments\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "API_KEY = \"AIzaSyC14K1XMc5q5wOmHKzuLH5zWb_BED8SrkQ\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Car.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TO-DO\n",
    "- A single run-through an experiment should work now\n",
    "    - Troubleshoot preprocessing method\n",
    "    - Clarify benchmarking method (or how we pass data to it) so that it's taken on out-of-sample data    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([232])) that is different to the input size (torch.Size([232, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 49.8359\n",
      "Epoch [20/100], Loss: 48.4358\n",
      "Epoch [30/100], Loss: 47.0938\n",
      "Epoch [40/100], Loss: 45.8054\n",
      "Epoch [50/100], Loss: 44.5666\n",
      "Epoch [60/100], Loss: 43.3743\n",
      "Epoch [70/100], Loss: 42.2254\n",
      "Epoch [80/100], Loss: 41.1180\n",
      "Epoch [90/100], Loss: 40.0507\n",
      "Epoch [100/100], Loss: 39.0220\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([232])) that is different to the input size (torch.Size([232, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 46.5554\n",
      "Epoch [20/100], Loss: 45.2375\n",
      "Epoch [30/100], Loss: 43.9608\n",
      "Epoch [40/100], Loss: 42.7220\n",
      "Epoch [50/100], Loss: 41.5184\n",
      "Epoch [60/100], Loss: 40.3482\n",
      "Epoch [70/100], Loss: 39.2103\n",
      "Epoch [80/100], Loss: 38.1041\n",
      "Epoch [90/100], Loss: 37.0301\n",
      "Epoch [100/100], Loss: 35.9897\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([234])) that is different to the input size (torch.Size([234, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 46.0803\n",
      "Epoch [20/100], Loss: 44.7571\n",
      "Epoch [30/100], Loss: 43.4690\n",
      "Epoch [40/100], Loss: 42.2121\n",
      "Epoch [50/100], Loss: 40.9831\n",
      "Epoch [60/100], Loss: 39.7801\n",
      "Epoch [70/100], Loss: 38.6024\n",
      "Epoch [80/100], Loss: 37.4499\n",
      "Epoch [90/100], Loss: 36.3232\n",
      "Epoch [100/100], Loss: 35.2247\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([232])) that is different to the input size (torch.Size([232, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 49.3818\n",
      "Epoch [20/100], Loss: 47.8391\n",
      "Epoch [30/100], Loss: 46.3522\n",
      "Epoch [40/100], Loss: 44.9172\n",
      "Epoch [50/100], Loss: 43.5307\n",
      "Epoch [60/100], Loss: 42.1906\n",
      "Epoch [70/100], Loss: 40.8950\n",
      "Epoch [80/100], Loss: 39.6433\n",
      "Epoch [90/100], Loss: 38.4360\n",
      "Epoch [100/100], Loss: 37.2744\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([234])) that is different to the input size (torch.Size([234, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 49.9341\n",
      "Epoch [20/100], Loss: 48.3429\n",
      "Epoch [30/100], Loss: 46.8116\n",
      "Epoch [40/100], Loss: 45.3349\n",
      "Epoch [50/100], Loss: 43.9087\n",
      "Epoch [60/100], Loss: 42.5297\n",
      "Epoch [70/100], Loss: 41.1956\n",
      "Epoch [80/100], Loss: 39.9045\n",
      "Epoch [90/100], Loss: 38.6561\n",
      "Epoch [100/100], Loss: 37.4503\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([234])) that is different to the input size (torch.Size([234, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 43.3315\n",
      "Epoch [20/100], Loss: 42.0591\n",
      "Epoch [30/100], Loss: 40.8270\n",
      "Epoch [40/100], Loss: 39.6326\n",
      "Epoch [50/100], Loss: 38.4739\n",
      "Epoch [60/100], Loss: 37.3498\n",
      "Epoch [70/100], Loss: 36.2593\n",
      "Epoch [80/100], Loss: 35.2032\n",
      "Epoch [90/100], Loss: 34.1822\n",
      "Epoch [100/100], Loss: 33.1971\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([232])) that is different to the input size (torch.Size([232, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 48.0536\n",
      "Epoch [20/100], Loss: 46.7890\n",
      "Epoch [30/100], Loss: 45.5680\n",
      "Epoch [40/100], Loss: 44.3871\n",
      "Epoch [50/100], Loss: 43.2418\n",
      "Epoch [60/100], Loss: 42.1290\n",
      "Epoch [70/100], Loss: 41.0464\n",
      "Epoch [80/100], Loss: 39.9924\n",
      "Epoch [90/100], Loss: 38.9666\n",
      "Epoch [100/100], Loss: 37.9682\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([234])) that is different to the input size (torch.Size([234, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 39.2844\n",
      "Epoch [20/100], Loss: 37.9955\n",
      "Epoch [30/100], Loss: 36.7513\n",
      "Epoch [40/100], Loss: 35.5482\n",
      "Epoch [50/100], Loss: 34.3827\n",
      "Epoch [60/100], Loss: 33.2520\n",
      "Epoch [70/100], Loss: 32.1542\n",
      "Epoch [80/100], Loss: 31.0877\n",
      "Epoch [90/100], Loss: 30.0513\n",
      "Epoch [100/100], Loss: 29.0448\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([234])) that is different to the input size (torch.Size([234, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 48.2062\n",
      "Epoch [20/100], Loss: 46.8786\n",
      "Epoch [30/100], Loss: 45.5977\n",
      "Epoch [40/100], Loss: 44.3600\n",
      "Epoch [50/100], Loss: 43.1628\n",
      "Epoch [60/100], Loss: 42.0032\n",
      "Epoch [70/100], Loss: 40.8790\n",
      "Epoch [80/100], Loss: 39.7892\n",
      "Epoch [90/100], Loss: 38.7338\n",
      "Epoch [100/100], Loss: 37.7130\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.01\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([232])) that is different to the input size (torch.Size([232, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 45.2886\n",
      "Epoch [20/100], Loss: 44.0865\n",
      "Epoch [30/100], Loss: 42.9146\n",
      "Epoch [40/100], Loss: 41.7719\n",
      "Epoch [50/100], Loss: 40.6578\n",
      "Epoch [60/100], Loss: 39.5720\n",
      "Epoch [70/100], Loss: 38.5148\n",
      "Epoch [80/100], Loss: 37.4875\n",
      "Epoch [90/100], Loss: 36.4920\n",
      "Epoch [100/100], Loss: 35.5302\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([338])) that is different to the input size (torch.Size([338, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 45.8377\n",
      "Epoch [20/100], Loss: 44.4479\n",
      "Epoch [30/100], Loss: 43.0974\n",
      "Epoch [40/100], Loss: 41.7831\n",
      "Epoch [50/100], Loss: 40.5024\n",
      "Epoch [60/100], Loss: 39.2533\n",
      "Epoch [70/100], Loss: 38.0343\n",
      "Epoch [80/100], Loss: 36.8440\n",
      "Epoch [90/100], Loss: 35.6822\n",
      "Epoch [100/100], Loss: 34.5495\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([327])) that is different to the input size (torch.Size([327, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 37.4114\n",
      "Epoch [20/100], Loss: 36.2625\n",
      "Epoch [30/100], Loss: 35.1523\n",
      "Epoch [40/100], Loss: 34.0775\n",
      "Epoch [50/100], Loss: 33.0345\n",
      "Epoch [60/100], Loss: 32.0208\n",
      "Epoch [70/100], Loss: 31.0350\n",
      "Epoch [80/100], Loss: 30.0754\n",
      "Epoch [90/100], Loss: 29.1408\n",
      "Epoch [100/100], Loss: 28.2309\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([303])) that is different to the input size (torch.Size([303, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 43.1414\n",
      "Epoch [20/100], Loss: 41.8656\n",
      "Epoch [30/100], Loss: 40.6239\n",
      "Epoch [40/100], Loss: 39.4129\n",
      "Epoch [50/100], Loss: 38.2302\n",
      "Epoch [60/100], Loss: 37.0733\n",
      "Epoch [70/100], Loss: 35.9409\n",
      "Epoch [80/100], Loss: 34.8315\n",
      "Epoch [90/100], Loss: 33.7447\n",
      "Epoch [100/100], Loss: 32.6804\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([315])) that is different to the input size (torch.Size([315, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 44.8980\n",
      "Epoch [20/100], Loss: 43.5623\n",
      "Epoch [30/100], Loss: 42.2608\n",
      "Epoch [40/100], Loss: 40.9912\n",
      "Epoch [50/100], Loss: 39.7519\n",
      "Epoch [60/100], Loss: 38.5420\n",
      "Epoch [70/100], Loss: 37.3605\n",
      "Epoch [80/100], Loss: 36.2071\n",
      "Epoch [90/100], Loss: 35.0827\n",
      "Epoch [100/100], Loss: 33.9887\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([327])) that is different to the input size (torch.Size([327, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 35.8067\n",
      "Epoch [20/100], Loss: 34.5010\n",
      "Epoch [30/100], Loss: 33.2466\n",
      "Epoch [40/100], Loss: 32.0390\n",
      "Epoch [50/100], Loss: 30.8747\n",
      "Epoch [60/100], Loss: 29.7507\n",
      "Epoch [70/100], Loss: 28.6647\n",
      "Epoch [80/100], Loss: 27.6149\n",
      "Epoch [90/100], Loss: 26.6003\n",
      "Epoch [100/100], Loss: 25.6206\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([325])) that is different to the input size (torch.Size([325, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 47.0979\n",
      "Epoch [20/100], Loss: 45.6297\n",
      "Epoch [30/100], Loss: 44.1998\n",
      "Epoch [40/100], Loss: 42.8048\n",
      "Epoch [50/100], Loss: 41.4421\n",
      "Epoch [60/100], Loss: 40.1097\n",
      "Epoch [70/100], Loss: 38.8061\n",
      "Epoch [80/100], Loss: 37.5312\n",
      "Epoch [90/100], Loss: 36.2846\n",
      "Epoch [100/100], Loss: 35.0675\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([320])) that is different to the input size (torch.Size([320, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 39.9591\n",
      "Epoch [20/100], Loss: 38.6613\n",
      "Epoch [30/100], Loss: 37.4032\n",
      "Epoch [40/100], Loss: 36.1811\n",
      "Epoch [50/100], Loss: 34.9922\n",
      "Epoch [60/100], Loss: 33.8338\n",
      "Epoch [70/100], Loss: 32.7047\n",
      "Epoch [80/100], Loss: 31.6039\n",
      "Epoch [90/100], Loss: 30.5313\n",
      "Epoch [100/100], Loss: 29.4871\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 2\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([320])) that is different to the input size (torch.Size([320, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 39.0491\n",
      "Epoch [20/100], Loss: 37.5917\n",
      "Epoch [30/100], Loss: 36.1843\n",
      "Epoch [40/100], Loss: 34.8223\n",
      "Epoch [50/100], Loss: 33.5021\n",
      "Epoch [60/100], Loss: 32.2212\n",
      "Epoch [70/100], Loss: 30.9772\n",
      "Epoch [80/100], Loss: 29.7692\n",
      "Epoch [90/100], Loss: 28.5964\n",
      "Epoch [100/100], Loss: 27.4600\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([301])) that is different to the input size (torch.Size([301, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 45.0365\n",
      "Epoch [20/100], Loss: 43.6742\n",
      "Epoch [30/100], Loss: 42.3448\n",
      "Epoch [40/100], Loss: 41.0469\n",
      "Epoch [50/100], Loss: 39.7787\n",
      "Epoch [60/100], Loss: 38.5400\n",
      "Epoch [70/100], Loss: 37.3312\n",
      "Epoch [80/100], Loss: 36.1540\n",
      "Epoch [90/100], Loss: 35.0102\n",
      "Epoch [100/100], Loss: 33.9025\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.1\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([320])) that is different to the input size (torch.Size([320, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 37.7583\n",
      "Epoch [20/100], Loss: 36.2207\n",
      "Epoch [30/100], Loss: 34.7582\n",
      "Epoch [40/100], Loss: 33.3651\n",
      "Epoch [50/100], Loss: 32.0361\n",
      "Epoch [60/100], Loss: 30.7667\n",
      "Epoch [70/100], Loss: 29.5530\n",
      "Epoch [80/100], Loss: 28.3921\n",
      "Epoch [90/100], Loss: 27.2819\n",
      "Epoch [100/100], Loss: 26.2212\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 0\n",
      "NO VALID JSON (AH)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/neural_network.py:40: RuntimeWarning: invalid value encountered in divide\n",
      "  inputs = (inputs - inputs.mean(axis=0)) / inputs.std(axis=0)\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([225])) that is different to the input size (torch.Size([225, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loss is NaN. Stopping training.\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([264])) that is different to the input size (torch.Size([264, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 41.2898\n",
      "Epoch [20/100], Loss: 39.9505\n",
      "Epoch [30/100], Loss: 38.6604\n",
      "Epoch [40/100], Loss: 37.4162\n",
      "Epoch [50/100], Loss: 36.2145\n",
      "Epoch [60/100], Loss: 35.0531\n",
      "Epoch [70/100], Loss: 33.9304\n",
      "Epoch [80/100], Loss: 32.8457\n",
      "Epoch [90/100], Loss: 31.7981\n",
      "Epoch [100/100], Loss: 30.7874\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 2\n",
      "NO VALID JSON (AH)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/neural_network.py:40: RuntimeWarning: invalid value encountered in divide\n",
      "  inputs = (inputs - inputs.mean(axis=0)) / inputs.std(axis=0)\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([225])) that is different to the input size (torch.Size([225, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loss is NaN. Stopping training.\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([345])) that is different to the input size (torch.Size([345, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 49.6514\n",
      "Epoch [20/100], Loss: 48.0755\n",
      "Epoch [30/100], Loss: 46.5523\n",
      "Epoch [40/100], Loss: 45.0765\n",
      "Epoch [50/100], Loss: 43.6441\n",
      "Epoch [60/100], Loss: 42.2514\n",
      "Epoch [70/100], Loss: 40.8957\n",
      "Epoch [80/100], Loss: 39.5744\n",
      "Epoch [90/100], Loss: 38.2869\n",
      "Epoch [100/100], Loss: 37.0321\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([279])) that is different to the input size (torch.Size([279, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 53.5192\n",
      "Epoch [20/100], Loss: 51.8564\n",
      "Epoch [30/100], Loss: 50.2392\n",
      "Epoch [40/100], Loss: 48.6646\n",
      "Epoch [50/100], Loss: 47.1303\n",
      "Epoch [60/100], Loss: 45.6343\n",
      "Epoch [70/100], Loss: 44.1760\n",
      "Epoch [80/100], Loss: 42.7557\n",
      "Epoch [90/100], Loss: 41.3747\n",
      "Epoch [100/100], Loss: 40.0353\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([262])) that is different to the input size (torch.Size([262, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 48.7685\n",
      "Epoch [20/100], Loss: 47.2517\n",
      "Epoch [30/100], Loss: 45.7782\n",
      "Epoch [40/100], Loss: 44.3433\n",
      "Epoch [50/100], Loss: 42.9436\n",
      "Epoch [60/100], Loss: 41.5760\n",
      "Epoch [70/100], Loss: 40.2385\n",
      "Epoch [80/100], Loss: 38.9307\n",
      "Epoch [90/100], Loss: 37.6516\n",
      "Epoch [100/100], Loss: 36.4027\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([249])) that is different to the input size (torch.Size([249, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 53.6945\n",
      "Epoch [20/100], Loss: 52.2549\n",
      "Epoch [30/100], Loss: 50.8601\n",
      "Epoch [40/100], Loss: 49.5054\n",
      "Epoch [50/100], Loss: 48.1865\n",
      "Epoch [60/100], Loss: 46.9001\n",
      "Epoch [70/100], Loss: 45.6435\n",
      "Epoch [80/100], Loss: 44.4143\n",
      "Epoch [90/100], Loss: 43.2109\n",
      "Epoch [100/100], Loss: 42.0322\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 2\n",
      "NO VALID JSON (AH)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([225])) that is different to the input size (torch.Size([225, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 48.3316\n",
      "Epoch [20/100], Loss: 46.9704\n",
      "Epoch [30/100], Loss: 45.6515\n",
      "Epoch [40/100], Loss: 44.3730\n",
      "Epoch [50/100], Loss: 43.1335\n",
      "Epoch [60/100], Loss: 41.9322\n",
      "Epoch [70/100], Loss: 40.7685\n",
      "Epoch [80/100], Loss: 39.6431\n",
      "Epoch [90/100], Loss: 38.5581\n",
      "Epoch [100/100], Loss: 37.5159\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 3\n",
      "NO VALID JSON (AH)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([225])) that is different to the input size (torch.Size([225, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 53.1250\n",
      "Epoch [20/100], Loss: 51.7265\n",
      "Epoch [30/100], Loss: 50.3615\n",
      "Epoch [40/100], Loss: 49.0268\n",
      "Epoch [50/100], Loss: 47.7196\n",
      "Epoch [60/100], Loss: 46.4384\n",
      "Epoch [70/100], Loss: 45.1820\n",
      "Epoch [80/100], Loss: 43.9505\n",
      "Epoch [90/100], Loss: 42.7443\n",
      "Epoch [100/100], Loss: 41.5645\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 0.5\n",
      "Drop indicator (0 is false): 1\n",
      "Trial # 4\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([264])) that is different to the input size (torch.Size([264, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch [10/100], Loss: 43.2836\n",
      "Epoch [20/100], Loss: 41.7997\n",
      "Epoch [30/100], Loss: 40.3543\n",
      "Epoch [40/100], Loss: 38.9446\n",
      "Epoch [50/100], Loss: 37.5688\n",
      "Epoch [60/100], Loss: 36.2259\n",
      "Epoch [70/100], Loss: 34.9159\n",
      "Epoch [80/100], Loss: 33.6396\n",
      "Epoch [90/100], Loss: 32.3980\n",
      "Epoch [100/100], Loss: 31.1942\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 1\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 0\n",
      "NO VALID JSON (AH)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/neural_network.py:40: RuntimeWarning: invalid value encountered in divide\n",
      "  inputs = (inputs - inputs.mean(axis=0)) / inputs.std(axis=0)\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([225])) that is different to the input size (torch.Size([225, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loss is NaN. Stopping training.\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 1\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 1\n",
      "NO VALID JSON (AH)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:44: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'skew': skew(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/experiment_runner.py:45: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.\n",
      "  'kurtosis': kurtosis(col_data),\n",
      "/Users/elizaknapp/Desktop/Harvard/am_226/am226/neural_network.py:40: RuntimeWarning: invalid value encountered in divide\n",
      "  inputs = (inputs - inputs.mean(axis=0)) / inputs.std(axis=0)\n",
      "/Users/elizaknapp/miniconda3/lib/python3.12/site-packages/torch/nn/modules/loss.py:608: UserWarning: Using a target size (torch.Size([225])) that is different to the input size (torch.Size([225, 1])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.\n",
      "  return F.mse_loss(input, target, reduction=self.reduction)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loss is NaN. Stopping training.\n",
      "Current experiment\n",
      "Subset size: 0.05\n",
      "Row percentage: 1\n",
      "Drop indicator (0 is false): 0\n",
      "Trial # 2\n"
     ]
    }
   ],
   "source": [
    "# Define the dataset\n",
    "dataset = pd.read_csv('car.csv') \n",
    "target_column = 'Selling_Price' \n",
    "\n",
    "# These are in fractions\n",
    "subset_sizes = [0.05, 0.1] # What happens when you double the number of examples given\n",
    "# Number of rows to generate\n",
    "row_sizes =  [0.01, 0.1, 0.5, 1] # This should be len(dataset) * array number\n",
    "# Number of experiments to run\n",
    "num_trials = 5\n",
    "\n",
    "# Initialize synthetic data generator and experiment runner\n",
    "synthetic_data_generator = GenerateSyntheticData(API_KEY)\n",
    "# Initialize the preprocessor based on the given dataset\n",
    "preprocessor = PreprocessData(dataset, target_column)\n",
    "\n",
    "# Loop through subset sizes\n",
    "results = []\n",
    "for prop in subset_sizes:\n",
    "    for n in row_sizes:\n",
    "        for s in range(2): # This is whether we drop indicator\n",
    "            for i in range(num_trials): \n",
    "                print(\"Current experiment\")\n",
    "                print(f\"Subset size: {prop}\")\n",
    "                print(f\"Row percentage: {n}\")\n",
    "                print(f\"Drop indicator (0 is false): {s}\")\n",
    "                print(f\"Trial # {i}\")\n",
    "                # Take subset of the data\n",
    "                subset = dataset.sample(frac=prop)\n",
    "                # Generate synthetic data using subset + additional information\n",
    "                synthetic_data = synthetic_data_generator.predict(n,len(dataset),subset)\n",
    "                # Combine real and synthetic data for training\n",
    "                dataset['source'] = 0\n",
    "                synthetic_data['source'] = 1\n",
    "\n",
    "                # Create a train test split \n",
    "                train_data, test_data = train_test_split(dataset)\n",
    "                combined_data = pd.concat([train_data, synthetic_data])\n",
    "\n",
    "                # Preprocess data\n",
    "                combined_df_processed = preprocessor.preprocess(combined_data)\n",
    "                test_data_processed = preprocessor.preprocess(test_data)\n",
    "                \n",
    "                if s != 0:\n",
    "                    # If we don't check source column\n",
    "                    combined_df_processed = combined_df_processed.drop(columns=['source'])\n",
    "                    test_data_processed = test_data_processed.drop(columns=['source'])\n",
    "\n",
    "                # Initialize an experiment runner\n",
    "                experiment_runner = ExperimentRunner(combined_df_processed, target_column)\n",
    "                # Compute subset characteristics (dimensions, variance, skewness)\n",
    "                subset_characteristics = experiment_runner.compute_characteristics(subset)\n",
    "                # Compute generated data characteristics\n",
    "                generated_characteristics = experiment_runner.compute_characteristics(synthetic_data)\n",
    "                # Train the network and benchmark\n",
    "                success = experiment_runner.train_network(combined_df_processed)\n",
    "                if not success:\n",
    "                    mse = \"Trial failed because gradient blowup\"\n",
    "                else:\n",
    "                    # Benchmark network\n",
    "                    mse = experiment_runner.benchmark_network(test_data_processed)\n",
    "                    # Save results\n",
    "                    results.append({\n",
    "                        \"source\": prop,\n",
    "                        \"generated_rows\" : n,\n",
    "                        \"subset_id\" : i,\n",
    "                        \"indicators\" : s,\n",
    "                        \"target_column\": target_column,\n",
    "                        \"subset_characteristics\": subset_characteristics.to_dict(),\n",
    "                        \"generated_characteristics\" : generated_characteristics.to_dict(),\n",
    "                        \"mse\": mse\n",
    "                    })\n",
    "\n",
    "# Write results to a json\n",
    "file_path = \"results.json\"\n",
    "\n",
    "# Write the list of dictionaries to a JSON file\n",
    "with open(file_path, \"w\") as json_file:\n",
    "    json.dump(results, json_file, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "for result in results:\n",
    "    result[\"subset_characteristics\"] = result[\"subset_characteristics\"].to_dict()\n",
    "    result[\"generated_characteristics\"] = result[\"generated_characteristics\"].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the file path\n",
    "file_path = \"results.json\"\n",
    "\n",
    "# Write the list of dictionaries to a JSON file\n",
    "with open(file_path, \"w\") as json_file:\n",
    "    json.dump(results, json_file, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.True_"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "False ==combined_df_processed.isna().any().any()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ideas for visualizations\n",
    "- GENERAL IDEA: See if there's a relationship between proportion of \"realness\" ((prop * len(df)) / (n + prop * len(df))) and preservation of characteristics (average of difference between subset characteristics and generated characteristics) and mse\n",
    "    - Are there certain characteristics (mean, variance, etc.) that are preserved better on average by feature? What properties about the feature make it the case?\n",
    "- Coolest graph would be something like proportion of \"realness\" on x-axis and mse and then two lines corresponding to with and without-indicators\n",
    "    - Could be super cool to see something besides \"indicators always beat out without indicators\" - something unintuitive would be sick\n",
    "- Write down any other ideas!\n",
    "- Just a bunch of facet grids could be cool"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Indicator ideas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Weights on how much the model uses the indicators as information\n",
    "- also try with and without indicators and compare performance\n",
    "- maybe we can compare embedding care "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
